components/renku_data_services/session/shipwright_core.py

```diff
@@ -0,0 +1,111 @@
+"""Business logic for ShipWright resources."""
+
+import logging
+from typing import TYPE_CHECKING
+
+from renku_data_services.session import constants, crs, models
+from renku_data_services.session.k8s_client import ShipwrightClient
+
+if TYPE_CHECKING:
+    from renku_data_services.app_config.config import BuildsConfig
+
+
+async def update_build_status(
+    build: models.Build, shipwright_client: ShipwrightClient | None
+) -> models.ShipWrightBuildStatusUpdate | None:
+    """Update the status of a build by pulling the corresponding BuildRun from ShipWright.
+
+    Note: this method will update `build` in place.
+    """
+    if shipwright_client is None:
+        logging.warning("ShipWright client not defined, BuildRun refresh skipped.")
+        return None
+
+    k8s_name = build.get_k8s_name()
+    k8s_build = await shipwright_client.get_build_run(name=k8s_name)
+
+    if k8s_build is None:
+        return models.ShipWrightBuildStatusUpdate(status=models.BuildStatus.failed)
```

Comment:
question: I find it confusing that if the client is not defined we return None. But if we cannot find the build run we set the status to failed. Why not return None in this case too. Then somewhere else decide what the status should be when you cannot find the build run.

Reply:
`None` means "no update". If there is no underlying k8s to query from, then there is no update to consider. Note that this is so that tests run without creating buildruns.

Reply:
Done in [705fef2](https://github.com/SwissDataScienceCenter/renku-data-services/pull/625/commits/705fef2a1e8db539f9678d57f309ceb8ecc9eff9).